pkg/controller/job/pod_failure_policy.go

```diff
@@ -0,0 +1,116 @@
+	if podFailurePolicy != nil {
+		for index, podFailurePolicyRule := range podFailurePolicy.Rules {
+			if podFailurePolicyRule.OnExitCodes != nil {
+				if containerStatus := matchOnExitCodes(&failedPod.Status, podFailurePolicyRule.OnExitCodes); containerStatus != nil {
+					switch podFailurePolicyRule.Action {
+					case batch.PodFailurePolicyActionIgnore:
+						return nil, false
+					case batch.PodFailurePolicyActionCount:
+						return nil, true
+					case batch.PodFailurePolicyActionFailJob:
+						msg := fmt.Sprintf("Container %s for pod %s/%s failed with exit code %v matching %v rule at index %d",
+							containerStatus.Name, failedPod.Namespace, failedPod.Name, containerStatus.State.Terminated.ExitCode, podFailurePolicyRule.Action, index)
+						return &msg, true
+					}
+				}
+			} else if podFailurePolicyRule.OnPodConditions != nil {
+				if podCondition := matchOnPodConditions(&failedPod.Status, podFailurePolicyRule.OnPodConditions); podCondition != nil {
+					switch podFailurePolicyRule.Action {
+					case batch.PodFailurePolicyActionIgnore:
+						return nil, false
+					case batch.PodFailurePolicyActionCount:
+						return nil, true
+					case batch.PodFailurePolicyActionFailJob:
+						msg := fmt.Sprintf("Pod %s/%s has condition %v matching %v rule at index %d",
+							failedPod.Namespace, failedPod.Name, podCondition.Type, podFailurePolicyRule.Action, index)
+						return &msg, true
+					}
+				}
+			}
+		}
+	}
+	return nil, true
+}
+
+func matchOnExitCodes(podStatus *v1.PodStatus, requirement *batch.PodFailurePolicyOnExitCodesRequirement) *v1.ContainerStatus {
+	if containerStatus := getMatchingContainerFromList(podStatus.ContainerStatuses, requirement); containerStatus != nil {
+		return containerStatus
+	}
+	return getMatchingContainerFromList(podStatus.InitContainerStatuses, requirement)
+}
+
+func matchOnPodConditions(podStatus *v1.PodStatus, requirement []batch.PodFailurePolicyOnPodConditionsPattern) *v1.PodCondition {
+	for _, podCondition := range podStatus.Conditions {
+		for _, pattern := range requirement {
+			if podCondition.Type == pattern.Type && podCondition.Status == v1.ConditionTrue {
+				return &podCondition
+			}
+		}
+	}
+	return nil
+}
+
+func getMatchingContainerFromList(containerStatuses []v1.ContainerStatus, requirement *batch.PodFailurePolicyOnExitCodesRequirement) *v1.ContainerStatus {
+	for _, containerStatus := range containerStatuses {
+		if requirement.ContainerName == nil || *requirement.ContainerName == containerStatus.Name {
+			if containerStatus.State.Terminated.ExitCode != 0 {
+				if isOnExitCodesOperatorMatching(containerStatus.State.Terminated.ExitCode, requirement.Operator, requirement.Values) {
+					return &containerStatus
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func isOnExitCodesOperatorMatching(exitCode int32, operator batch.PodFailurePolicyOnExitCodesOperator, values []int32) bool {
```

Comment:
Nit: seems like it's easier to pass entire `batch.PodFailurePolicyOnExitCodesRequirement` rather than operator and values separately.

Reply:
updated